subsys/net/l2/ethernet/bridge.c

```diff
@@ -165,79 +220,247 @@ static inline bool is_link_local_addr(struct net_eth_addr *addr)
+	ctx->is_setup = false;
+}
+
+static enum virtual_interface_caps bridge_get_capabilities(struct net_if *iface)
+{
+	ARG_UNUSED(iface);
+
+	return VIRTUAL_INTERFACE_BRIDGE;
+}
+
+static int bridge_iface_start(const struct device *dev)
+{
+	struct eth_bridge_iface_context *ctx = dev->data;
+
+	if (!ctx->is_setup) {
+		NET_DBG("Bridge interface %d not configured yet.",
+			net_if_get_by_iface(ctx->iface));
+		return -ENOENT;
+	}
+
+	if (ctx->status) {
+		return -EALREADY;
+	}
+
+	ctx->status = true;
+
+	NET_DBG("Starting iface %d", net_if_get_by_iface(ctx->iface));
+
+	/* You can implement here any special action that is needed
+	 * when the network interface is coming up.
+	 */
+
+	NET_INFO("Bridge %d is %sactive", net_if_get_by_iface(ctx->iface), "");
+
+	net_virtual_set_name(ctx->iface, "<enabled>");
+
+	return 0;
+}
+
+static int bridge_iface_stop(const struct device *dev)
+{
+	struct eth_bridge_iface_context *ctx = dev->data;
+
+	if (!ctx->status) {
+		return -EALREADY;
+	}
+
+	ctx->status = false;
+
+	NET_DBG("Stopping iface %d", net_if_get_by_iface(ctx->iface));
+
+	/* You can implement here any special action that is needed
+	 * when the network interface is going down.
+	 */
+
+	NET_INFO("Bridge %d is %sactive", net_if_get_by_iface(ctx->iface), "not ");
+
+	if (ctx->is_setup) {
+		net_virtual_set_name(ctx->iface, "<disabled>");
+	} else {
+		net_virtual_set_name(ctx->iface, "<no config>");
+	}
+
+	return 0;
+}
+
+static enum net_verdict bridge_iface_process(struct net_if *iface,
+					     struct net_pkt *pkt,
+					     bool is_send)
+{
+	struct eth_bridge_iface_context *ctx = net_if_get_device(iface)->data;
+	struct net_if *orig_iface;
+	struct net_pkt *send_pkt;
+	size_t count;
 
 	/* Drop all link-local packets for now. */
 	if (is_link_local_addr((struct net_eth_addr *)net_pkt_lladdr_dst(pkt))) {
+		NET_DBG("DROP: lladdr");
 		return NET_DROP;
 	}
 
-	lock_bridge(br);
+	lock_bridge(ctx);
 
-	/*
-	 * Send packet to all registered interfaces for now.
-	 * Eventually we could get smarter with a MAC address cache.
-	 */
-	SYS_SLIST_FOR_EACH_NODE(&br->interfaces, node) {
-		struct ethernet_context *out_ctx;
-		struct net_pkt *out_pkt;
+	/* Keep the original packet so that we can send to each bridged interface */
+	orig_iface = net_pkt_orig_iface(pkt);
 
-		out_ctx = CONTAINER_OF(node, struct ethernet_context, bridge.node);
+	count = ctx->count;
 
-		/* Don't xmit on the same interface as the incoming packet's */
-		if (ctx == out_ctx) {
-			continue;
+	/* Pass the data to all the Ethernet interface except the originator
+	 * Ethernet interface.
+	 */
+	ARRAY_FOR_EACH(ctx->eth_iface, i) {
+		if (ctx->eth_iface[i] != NULL && ctx->eth_iface[i] != orig_iface &&
+		    ctx->iface != iface) {
+			/* Skip it if not up */
+			if (!net_if_flag_is_set(ctx->eth_iface[i], NET_IF_UP)) {
+				continue;
+			}
+
+			/* Clone the packet if we have more than two interfaces in the bridge
+			 * because the first send might mess the data part of the message.
```

Comment:
> because the first send might mess the data part of the message.

I don't think I follow this explanation, can you elaborate?

Reply:
The first fragment is removed in `ethernet_send()` after sending, so we need to do a full net_pkt clone, otherwise the data gets lost.
https://github.com/zephyrproject-rtos/zephyr/blob/0cd3f56f0d1acee632cdae0c5fcb2b6d19e80e8a/subsys/net/l2/ethernet/ethernet.c#L727

Reply:
So if we have only two Ethernet interfaces in the bridge, the bridged packet needs to be sent to only one Ethernet interface so we do not need to clone the packet.